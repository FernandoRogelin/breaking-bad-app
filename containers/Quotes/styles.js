import styled, { css } from 'styled-components';

import ButtonComponent from '#/components/Button';
import ContentComponent from '#/components/Content';

export const Wrapper = styled.div`
  display: flex;
  align-items: center;
  flex-direction: column;
`;

export const Title = styled.h1`
  margin: 0;
  font-size: 1.5em;
`;

export const Top = styled.div`
  display: flex;
  padding: 8px 114px 8px 30px;
  justify-content: space-between;
  border-bottom: 1px solid #919191;
`;

export const Itens = styled.div`
  margin: 0 15px;
  padding: 6px 15px;

  &:not(:last-child) {
    border-bottom: solid 1px #ddd;
  }

  @media only screen and (max-width: 440px) {
    flex-direction: column;
  }
`;

export const Name = styled.p`
  margin: 0;
  color: #313944;
  font-size: 20px;
`;

export const Content = styled(ContentComponent)``;

export const Text = styled.p`
  margin: 0 0 0 10px;
`;

export const Info = styled.span`
  color: #8c9aa3;
  margin: 0 0 0 5px;

  &:not(:last-child) {
    margin-bottom: 10px;
  }
`;

export const Quote = styled.p`
  color: #8c9aa3;
  margin: 0 0 0 30px;

  &:not(:last-child) {
    margin-bottom: 10px;
  }
`;

export const Header = styled.header`
  display: flex;
  position: relative;
  align-items: center;
  margin-bottom: 26px;
`;

export const Button = styled(ButtonComponent)`
  font-size: 12px;
  padding: 5px 10px;

  ${({ top }) =>
    top &&
    css`
      left: -80px;
      padding: 1px 6px;
      position: absolute;
    `}
`;
