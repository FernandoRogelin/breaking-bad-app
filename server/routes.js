const routes = require('next-routes');

module.exports = routes()
  .add('login')
  .add('episodes')
  .add('characters')
  .add('quotes');
