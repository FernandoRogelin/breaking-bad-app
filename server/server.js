const Koa = require('koa');
const Router = require('koa-router');
const helmet = require('koa-helmet');
const logger = require('koa-logger');
const bodyParcer = require('koa-bodyparser');
const nextjs = require('next');

const nextRoutes = require('./routes');
const port = process.env.PORT || 8080;
const app = nextjs({ dev: process.env.NODE_ENV !== 'production' });
const nextHandler = nextRoutes.getRequestHandler(app);

app
  .prepare()
  .then(() => {
    const server = new Koa();
    const router = new Router();

    server
      .use(logger())
      .use(bodyParcer())
      .use(helmet());

    router.get('/', ctx => {
      ctx.redirect('/login');
    });

    // Handle NextJS pages
    router.get('*', async ctx => {
      await nextHandler(ctx.req, ctx.res);

      ctx.status = 200;
      ctx.respond = false;
    });

    server.use(router.routes(), router.allowedMethods());

    server.listen(port, err => {
      if (err) console.log(err);
      console.log(`Server running in port http://localhost:${port}`);
    });
  })
  .catch(ex => {
    console.log(ex.stack);
    process.exit(1);
  });
