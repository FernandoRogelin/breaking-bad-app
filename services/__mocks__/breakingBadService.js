export const getCharacters = jest.fn().mockResolvedValue({
  status: 200,
  data: [
    {
      appearance: [1, 2, 3, 4, 5],
      birthday: '09-07-1958',
      char_id: 1,
      img:
        'https://images.amcnetworks.com/amc.com/wp-content/uploads/2015/04/cast_bb_700x1000_walter-white-lg.jpg',
      name: 'Walter White',
      nickname: 'Heisenberg',
      occupation: ['High School Chemistry Teacher', 'Meth King Pin'],
      portrayed: 'Bryan Cranston',
      status: 'Presumed dead',
    },
    {
      appearance: [1, 2, 3, 4, 5],
      birthday: '09-24-1984',
      char_id: 2,
      img:
        'https://upload.wikimedia.org/wikipedia/en/thumb/f/f2/Jesse_Pinkman2.jpg/220px-Jesse_Pinkman2.jpg',
      name: 'Jesse Pinkman',
      nickname: "Cap n' Cook",
      occupation: ['Meth Dealer'],
      portrayed: 'Aaron Paul',
      status: 'Alive',
    },
    {
      appearance: [1, 2, 3, 4, 5],
      birthday: '08-11-1970',
      char_id: 3,
      img:
        'https://s-i.huffpost.com/gen/1317262/images/o-ANNA-GUNN-facebook.jpg',
      name: 'Skyler White',
      nickname: 'Sky',
      occupation: [
        'House wife',
        'Book Keeper',
        'Car Wash Manager',
        'Taxi Dispatcher',
      ],
      portrayed: 'Anna Gunn',
      status: 'Alive',
    },
    {
      appearance: [1, 2, 3, 4, 5],
      birthday: '07-08-1993',
      char_id: 4,
      img:
        'https://media1.popsugar-assets.com/files/thumbor/WeLUSvbAMS_GL4iELYAUzu7Bpv0/fit-in/1024x1024/filters:format_auto-!!-:strip_icc-!!-/2018/01/12/910/n/1922283/fb758e62b5daf3c9_TCDBRBA_EC011/i/RJ-Mitte-Walter-White-Jr.jpg',
      name: 'Walter White Jr.',
      nickname: 'Flynn',
      occupation: ['Teenager'],
      portrayed: 'RJ Mitte',
      status: 'Alive',
    },
    {
      appearance: [1, 2, 3, 4, 5],
      birthday: 'Unknown',
      char_id: 5,
      img:
        'https://upload.wikimedia.org/wikipedia/en/thumb/c/c1/Hank_Schrader2.jpg/220px-Hank_Schrader2.jpg',
      name: 'Henry Schrader',
      nickname: 'Hank',
      occupation: ['DEA Agent'],
      portrayed: 'Dean Norris',
      status: 'Deceased',
    },
  ],
});

export const getCharacterDetails = jest.fn().mockResolvedValue({
  status: 200,
  data: {
    appearance: '1,2,3',
    birthday: null,
    char_id: 31,
    img:
      'https://vignette.wikia.nocookie.net/breakingbad/images/c/c9/AdamPinkman.png/revision/latest?cb=20141117180836&path-prefix=es',
    name: 'Adam Pinkman',
    nickname: "Jesse's dad",
    occupation: 'unknown',
    portrayed: 'Michael Bofshever',
    status: 'Alive',
  },
});

export const getEpisodesList = jest.fn().mockResolvedValue({
  status: 200,
  data: [
    {
      air_date: '01-20-2008',
      characters: [
        'Walter White',
        'Jesse Pinkman',
        'Skyler White',
        'Hank Schrader',
        'Marie Schrader',
        'White White Jr.',
      ],
      episode: '1',
      episode_id: 1,
      season: '1',
      title: 'Pilot',
    },
    {
      air_date: '01-27-2008',
      characters: [
        'Walter White',
        'Jesse Pinkman',
        'Skyler White',
        'White White Jr.',
        'Krazy-8',
      ],
      episode: '2',
      episode_id: 2,
      season: '1',
      title: "Cat's in the Bag...",
    },
    {
      air_date: '02-10-2008',
      characters: [
        'Walter White',
        'Jesse Pinkman',
        'Skyler White',
        'Hank Schrader',
        'Marie Schrader',
        'White White Jr.',
      ],
      episode: '3',
      episode_id: 3,
      season: '1',
      title: "...And the Bag's in the River",
    },
    {
      air_date: '02-17-2008',
      characters: [
        'Walter White',
        'Jesse Pinkman',
        'Skyler White',
        'Hank Schrader',
        'Marie Schrader',
        'White White Jr.',
      ],
      episode: '4',
      episode_id: 4,
      season: '1',
      title: 'Cancer Man',
    },
  ],
});

export const getEpisodeDetails = jest.fn().mockResolvedValue({
  status: 200,
  data: {
    air_date: '05-03-2009',
    characters:
      'Walter White, Jesse Pinkman, Skyler White, Hank Schrader, Marie Schrader, White White Jr., Saul Goodman, Jane Margolis',
    episode: '9',
    episode_id: 16,
    season: '2',
    title: '4 Days Out',
  },
});

export const getQuotes = jest.fn().mockResolvedValue({
  status: 200,
  data: {},
});
