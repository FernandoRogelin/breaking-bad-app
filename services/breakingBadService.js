import axios from 'axios';

export const urlBase = 'https://www.breakingbadapi.com/api/';

export const getCharacters = () => axios.get(`${urlBase}characters`);

export const getCharacterDetails = id =>
  axios.get(`${urlBase}characters/${id}`);

export const getEpisodesList = () => axios.get(`${urlBase}episodes`);

export const getEpisodeDetails = id => axios.get(`${urlBase}episodes/${id}`);

export const getQuotes = () => axios.get(`${urlBase}quotes`);
