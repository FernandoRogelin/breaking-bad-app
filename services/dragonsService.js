import axios from 'axios';

export const urlBase =
  'http://5c4b2a47aa8ee500142b4887.mockapi.io/api/v1/dragon';

export const getDragons = () => axios.get(urlBase);

export const getDetailsDragon = id => axios.get(`${urlBase}/${id}`);

export const createDragon = ({ name, type, createdAt }) =>
  axios.post(urlBase, {
    name,
    type,
    createdAt,
  });

export const deleteDragon = id => axios.delete(`${urlBase}/${id}`);

export const editDragon = (id, { name, type, createdAt }) =>
  axios.put(`${urlBase}/${id}`, {
    name,
    type,
    createdAt,
  });
