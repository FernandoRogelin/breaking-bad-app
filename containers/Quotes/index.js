import React, { useState, useEffect } from 'react';

import { Router } from '#/server/routes';
import { getQuotes } from '#/services/breakingBadService';

import {
  Top,
  Name,
  Text,
  Title,
  Itens,
  Quote,
  Header,
  Button,
  Wrapper,
  Content,
} from './styles';

export default function Quotes() {
  const [quotes, setQuotes] = useState([]);
  const [fetched, setFetched] = useState(false);

  async function handleQuotesList() {
    try {
      const { data } = await getQuotes();

      setFetched(true);
      setQuotes(data);
    } catch (e) {
      console.log(e);
    }
  }

  useEffect(() => {
    if (!fetched) {
      handleQuotesList();
    }
  }, [fetched]);

  return (
    <Wrapper>
      <Header>
        <Button
          top
          data-testid="handle-characters-page"
          onClick={() => Router.pushRoute('characters')}
        >
          Return
        </Button>
        <Title>Quotes List</Title>
      </Header>
      {quotes && (
        <Content data-testid="quotes-list">
          <Top>
            <Text>Quotes</Text>
          </Top>
          {quotes
            .sort((a, b) =>
              a.author < b.author ? -1 : a.author > b.author ? 1 : 0
            )
            .map(({ author, quote, quote_id }) => (
              <Itens key={quote_id}>
                <Name>{author}</Name>
                <Quote>{quote}</Quote>
              </Itens>
            ))}
        </Content>
      )}
    </Wrapper>
  );
}
