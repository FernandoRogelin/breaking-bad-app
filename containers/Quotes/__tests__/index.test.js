import React from 'react';
import { render, wait, fireEvent } from 'react-testing-library';
import { Router } from '#/server/routes';

Router.pushRoute = jest.fn();

import { getQuotes } from '#/services/breakingBadService';
import Quotes from '..';

jest.mock('#/services/breakingBadService');

describe('Quotes container', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  test('Quotes snapshot', () => {
    const { container } = render(<Quotes />);

    expect(container).toMatchSnapshot();
  });

  test('Inicial flow', async () => {
    render(<Quotes />);

    await wait(() => expect(getQuotes).toHaveBeenCalledTimes(1));
  });

  test('Handle character page', () => {
    const { getByTestId } = render(<Quotes />);

    fireEvent.click(getByTestId('handle-characters-page'));
  });
});
