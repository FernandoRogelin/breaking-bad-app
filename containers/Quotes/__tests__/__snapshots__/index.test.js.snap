// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quotes container Quotes snapshot 1`] = `
<div>
  <div
    class="styles__Wrapper-w22mi7-0 hCttmQ"
  >
    <header
      class="styles__Header-w22mi7-9 ipfACt"
    >
      <button
        class="Button-n3cyj4-0 styles__Button-w22mi7-10 hqkQwQ"
        data-testid="handle-characters-page"
      >
        Return
      </button>
      <h1
        class="styles__Title-w22mi7-1 cPtDXM"
      >
        Quotes List
      </h1>
    </header>
    <div
      class="Content-vb879o-0 styles__Content-w22mi7-5 beZUzH"
      data-testid="quotes-list"
    >
      <div
        class="styles__Top-w22mi7-2 jLuHzA"
      >
        <p
          class="styles__Text-w22mi7-6 cCesFI"
        >
          Quotes
        </p>
      </div>
    </div>
  </div>
</div>
`;
